README.md
# [DashHD.js](https://github.com/dashhive/dashhd.js)

Manage HD Keys from HD Wallet Seed and Extended (xprv, xpub) Key Paths. \
(compatible with the [Hierarchical Deterministic Keys (BIP-32)][bip-32] spec)

> A fully-functional, production-ready reference implementation of Dash HD -
> suitable for learning DASH specs and protocols, and porting to other
> languages.

```text
HD Wallet Seed:  ac27495480225222079d7be181583751e86f571027b0497b5b5d11218e0a8a13332572917f0f8e5a589620c6f15b11c61dee327651a14c34e18231052e48c069
```

```text
HD Key Path:  m/44'/5'/0'/0/0
        WIF:  XCGKuZcKDjNhx8DaNKK4xwMMNzspaoToT6CafJAbBfQTi57buhLK
    Address:  XrZJJfEKRNobcuwWKTD3bDu8ou7XSWPbc9

HD Key Path:  m/44'/5'/1'/1/1
        WIF:  XF9murLtNpJaZXbwMxqJ6BhigEtu9NxfBCJDBokCJcqFkYkz3itz
    Address:  XueHW2ELMxoXzXcaHMxmwVWhcADE1W5s8c
```

(that's the canonical _Zeed_ seed, generated from the canonical _Zoomonic_)

[bip-32]: https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki
[bip-39]: https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki
[bip-44]: https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki
[dash-phrase]: https://github.com/dashhive/dashphrase.js

## Install

### Node, Bun, & Bundlers

```sh
npm install --save @dashincubator/secp256k1@1.x
npm install --save dashhd@3.x
```

```js
let DashHd = require("dashd");
```

### Browser

```html
<script src="https://unpkg.com/@dashincubator/secp256k1.js"></script>
<script src="https://unpkg.com/dashkeys@0.9/dashkeys.js"></script>
<script src="https://unpkg.com/dashhd@3.x/dashhd.js"></script>
```

## Usage

**example:**

```js
var DashHd = require("dashhd");
var seed =
  "a0c42a9c3ac6abf2ba6a9946ae83af18f51bf1c9fa7dacc4c92513cc4dd015834341c775dcd4c0fac73547c5662d81a9e9361a0aac604a73a321bd9103bce8af";
var hdkey = await DashHd.fromMasterSeed(Buffer.from(seed, "hex"));
console.log(hdkey.privateExtendedKey);

// => 'xprv9s21ZrQH143K2SKJK9EYRW3Vsg8tWVHRS54hAJasj1eGsQXeWDHLeuu5hpLHRbeKedDJM4Wj9wHHMmuhPF8dQ3bzyup6R7qmMQ1i1FtzNEW'

var xpub = await hdkey.getPublicExtendedKey();
console.log(xpub);
// => 'xpub661MyMwAqRbcEvPmRAmYndzERhyNux1GoHzHxgzVHMBFkCro3kbbCiDZZ5XabZDyXPj5mH3hktvkjhhUdCQxie5e1g4t2GuAWNbPmsSfDp2'
```

### `await DashHd.fromMasterSeed(seedBuffer[, versions])`

Creates an `hdkey` object from a master seed buffer. Accepts an optional
`versions` object.

```js
var seed =
  "a0c42a9c3ac6abf2ba6a9946ae83af18f51bf1c9fa7dacc4c92513cc4dd015834341c775dcd4c0fac73547c5662d81a9e9361a0aac604a73a321bd9103bce8af";
var hdkey = await DashHd.fromMasterSeed(Buffer.from(seed, "hex"));
```

### `await DashHd.fromExtendedKey(extendedKey[, versions, skipVerification])`

Creates an `hdkey` object from a `xprv` or `xpub` extended key string. Accepts
an optional `versions` object & an optional `skipVerification` boolean. If
`skipVerification` is set to true, then the provided public key's x (and y if
uncompressed) coordinate will not will be verified to be on the curve.

```js
var key =
  "xprvA2nrNbFZABcdryreWet9Ea4LvTJcGsqrMzxHx98MMrotbir7yrKCEXw7nadnHM8Dq38EGfSh6dqA9QWTyefMLEcBYJUuekgW4BYPJcr9E7j";
var hdkey = await DashHd.fromExtendedKey(key);
```

**or**

```js
var key =
  "xpub6FnCn6nSzZAw5Tw7cgR9bi15UV96gLZhjDstkXXxvCLsUXBGXPdSnLFbdpq8p9HmGsApME5hQTZ3emM2rnY5agb9rXpVGyy3bdW6EEgAtqt";
var hdkey = DashHd.fromExtendedKey(key);
```

---

### `await hdkey.derive(path)`

Derives the `hdkey` at `path` from the current `hdkey`.

```js
var seed =
  "fffcf9f6f3f0edeae7e4e1dedbd8d5d2cfccc9c6c3c0bdbab7b4b1aeaba8a5a29f9c999693908d8a8784817e7b7875726f6c696663605d5a5754514e4b484542";
var hdkey = await DashHd.fromMasterSeed(Buffer.from(seed, "hex"));
var childkey = await hdkey.derive("m/0/2147483647'/1");

var xprv = await childkey.getPrivateExtendedKey();
console.log(xprv);
// -> "xprv9zFnWC6h2cLgpmSA46vutJzBcfJ8yaJGg8cX1e5StJh45BBciYTRXSd25UEPVuesF9yog62tGAQtHjXajPPdbRCHuWS6T8XA2ECKADdw4Ef"

var xpub = await childkey.getPublicExtendedKey();
console.log(xpub);
// -> "xpub6DF8uhdarytz3FWdA8TvFSvvAh8dP3283MY7p2V4SeE2wyWmG5mg5EwVvmdMVCQcoNJxGoWaU9DCWh89LojfZ537wTfunKau47EL2dhHKon"
```

Newer, "hardened" derivation paths look like this:

```js
// as defined by BIP-44
var childkey = await hdkey.derive("m/44'/0'/0'/0/0");
```

### `hdkey.wipePrivateData()`

Wipes all record of the private key from the `hdkey` instance. After calling
this method, the instance will behave as if it was created via
`DashHd.fromExtendedKey(xpub)`.

### `hdkey.getPrivateKey()`

Get the `hdkey`'s private key, stored as a buffer.

### `hdkey.publicKey`

Get the `hdkey`'s public key, stored as a buffer.

### `await hdkey.getPrivateExtendedKey()`

Get the `hdkey`'s `xprv`, stored as a string.

### `await hdkey.getPublicExtendedKey()`

Get the `hdkey`'s `xpub`, stored as a string.

## References

- https://github.com/bitcoinjs/bitcoinjs-lib/blob/master/src/hdnode.js
- http://bip32.org/
- http://blog.richardkiss.com/?p=313
- https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki
- http://bitcoinmagazine.com/8396/deterministic-wallets-advantages-flaw/

## License

Copyright © 2023 Dash Incubator \
Copyright © 2023 AJ ONeal \
Copyright © 2018-2022 cryptocoinjs

MIT License


CHANGELOG.md
## 3.0.0 / 2023-02-03

- Hard Fork / Rebrand to DashHD.js
- Complete refactor to WebCrypto, and async/await, and $DASH Tools
- API redesign for consistency with $DASH Tools

## 2.1.0 / 2023-01-18

- Add `skipVerification` option to `HDKey.fromExtendedKey()` to allow skipping
  verification logic for performance
  ([#53](https://github.com/cryptocoinjs/hdkey/pull/53))
- Use `ripemd160` package; as Node v18+ `crypto.createHash()` does not support
  ripemd160 ([#51](https://github.com/cryptocoinjs/hdkey/pull/51))
- Fix `.sign()` method in some environments with polyfilled `Buffer`
  implementations ([#50](https://github.com/cryptocoinjs/hdkey/pull/50))
- Performance improvements
  ([#52](https://github.com/cryptocoinjs/hdkey/pull/52))

  2.0.1 / 2020-05-30

---

- Bugfix: prevent mutating buffers passed in
  ([#39](https://github.com/cryptocoinjs/hdkey/pull/39))

  2.0.0 / 2020-05-29

---

- **BREAKING:** Require Node.js v10+
  ([#38](https://github.com/cryptocoinjs/hdkey/pull/38))
- Upgrade `secp256k1` dependency; to use N-API
  ([#32](https://github.com/cryptocoinjs/hdkey/pull/32))

  1.1.2 / 2020-04-16

---

- Fix extremely rare types bug
  ([#33](https://github.com/cryptocoinjs/hdkey/pull/33))
- Use `bs58check` dependency instead of `coinstring`
  ([#30](https://github.com/cryptocoinjs/hdkey/pull/30))
- Don't publish test files
  ([#27](https://github.com/cryptocoinjs/hdkey/issues/27),
  [#34](https://github.com/cryptocoinjs/hdkey/pull/34))

  1.1.1 / 2019-02-09

---

- Fix Electron v4 support. No changes to external API.
  ([#26](https://github.com/cryptocoinjs/hdkey/pull/26))

  1.1.0 / 2018-08-14

---

- Add `wipePrivateData()` method
  ([#22](https://github.com/cryptocoinjs/hdkey/pull/22))
- Add missing LICENSE file
  ([#21](https://github.com/cryptocoinjs/hdkey/pull/21))

  1.0.0 / 2018-05-24

---

- drop support for all Node.js versions 4 and earlier
- fix `derive()` path validation
  ([#20](https://github.com/cryptocoinjs/hdkey/pull/20))

  0.8.0 / 2018-02-06

---

- add `sign()` and `verify()`
- upgrade to `safe-buffer`

  0.7.1 / 2016-05-26

---

- fix bug when `privateKey` is `null`, `privateExtendedKey` should not throw,
  and return `null` [#7][#7]

  0.7.0 / 2016-03-22

---

- upgrade from `ecurve` to `secp256k1`. [#5][#5]

  0.6.0 / 2015-07-02

---

- **breaking** (same day though, haha). Changed
  `publicExtendedKey`/`privateExtendedKey` in `JSON` methods to `xpub`/`xpriv`
- export `HARDENED_OFFSET`

  0.5.0 / 2015-07-02

---

- JavaScript Standard Style
- fix rare condition for BIP32 consistency: [#1][#1]
- added `toJSON()/fromJSON()`

  0.4.0 / 2014-09-24

---

- dropped `sha512` dependency and upgraded to crypto-browserify that supports
  sha512

  0.3.1 / 2014-07-11

---

- removed superfluous code `this._privateKeyBigInteger`

  0.3.0 / 2014-06-29

---

- bugfix: if private key was less than 32 bytes, pad out to 32 bytes with
  leading zeros (this happens in derive)
- changed behavior of `privateExtendedKey()` and `publicExtendedKey()` to return
  base 58 encoded `string` instead of `Buffer`
- changed behavior of `fromExtendedKey()` from accepting a type of `Buffer`
  bytes to base58 `string`

  0.2.0 / 2014-06-25

---

- upgraded `"ecurve": "^0.8.0"` to `"ecurve": "^1.0.0"`
- added functionality to derive public to public child keys

  0.1.0 / 2014-06-16

---

- removed semicolons per http://cryptocoinjs.com/about/contributing/#semicolons
- removed `ECKey` dep
- added `ecurve` dep
- removed `terst` dev dep for `assert`
- added method `fromMasterSeed(seedBuffer, [versions])`
- changed constructor from `new HDKey(masterSeed, [versions])` to
  `new HDKey([versions])`
- added properties: `privateKey` and `publicKey`
- removed method `getIdentifier()`, added property `identifier`
- removed method `getFingerprint()`, added property `fingerprint`
- renamed `private` to `privateExtendedKey`
- renamed `public` to `publicExtendedKey`
- added method `fromExtendedKey()`

  0.0.1 / 2014-05-29

---

- initial release

<!-- generated using jprichardson/issue-links -->

[#7]:
  https://github.com/cryptocoinjs/hdkey/issues/7
  "privateExtendedKey error handling"
[#6]:
  https://github.com/cryptocoinjs/hdkey/pull/6
  "hdkey: use bippath for BIP32 path parsing and validation"
[#5]:
  https://github.com/cryptocoinjs/hdkey/pull/5
  "hdkey: use the secp256k1 package for crypto"
[#4]:
  https://github.com/cryptocoinjs/hdkey/issues/4
  "Is this library still maintained?"
[#3]: https://github.com/cryptocoinjs/hdkey/pull/3 "Update hdkey.js"
[#2]: https://github.com/cryptocoinjs/hdkey/pull/2 "Update hdkey.js"
[#1]:
  https://github.com/cryptocoinjs/hdkey/issues/1
  "rare condition needed for bip consistency"


test.js
"use strict";

console.error("TODO: needs tests");
process.exit(1);


dashhd.js
/**
 * @typedef DashHD
 * @prop {HDCreate} create
 * @prop {HDFingerprint} fingerprint
 * @prop {HDFromSeed} fromMasterSeed
 * @prop {HDFromXKey} fromExtendedKey
 * @prop {HDUtils} utils
 * @prop {Number} HARDENED_OFFSET - 0x80000000
 */

/**
 * @typedef HDUtils
 * @prop {HDKeyTweak} privateKeyTweakAdd
 * @prop {HDKeyTweak} publicKeyTweakAdd
 * @prop {HDKeyToKey} publicKeyNormalize
 * @prop {HDHasher} ripemd160sum
 * @prop {HDHasher} sha256sum
 * @prop {HDHasher} sha512hmac
 * @prop {HDKeyToKey} toPublicKey
 */

/**
 * @callback HDCreate
 * @param {HDVersions} [versions]
 * @returns {HDKey}
 */

/**
 * @typedef HDKey
 * @prop {Uint8Array} chainCode - extra 32-bytes of shared entropy for xkeys
 * @prop {Number} depth - of hd path - typically 0 is seed, 1-3 hardened, 4-5 are not
 * @prop {Number} index - the final segment of an HD Path, the index of the wif/addr
 * @prop {Number} parentFingerprint - 32-bit int, slice of id, stored in child xkeys
 * @prop {Uint8Array} publicKey
 * @prop {HDVersions} versions - magic bytes for base58 prefix
 * @prop {HDDerivePath} derive - derive a full hd path from the given root
 * @prop {HDDeriveChild} deriveChild - get the next child xkey (in a path segment)
 * @prop {HDDeriveChild} _deriveChild - helper
 * @prop {HDMaybeGetString} getPrivateExtendedKey
 * @prop {HDMaybeGetBuffer} getPrivateKey
 * @prop {HDGetString} getPublicExtendedKey
 * @prop {HDSetBuffer} setPrivateKey
 * @prop {HDWipePrivates} wipePrivateData - randomizes private key buffer in-place
 */

/** @type {DashHD} */
//@ts-ignore
var DashHd = ("object" === typeof module && exports) || {};
(function (window, DashHd) {
  "use strict";

  //const BUFFER_LE = true;
  const BUFFER_BE = false;
  const COMPRESSED = true;

  //@ts-ignore
  let Crypto = window.crypto || require("crypto");

  //@ts-ignore
  /** @type {import('../node_modules/dashkeys/dashkeys.js')} */
  //@ts-ignore
  let DashKeys = window.DashKeys || require("dashkeys");

  let Utils = {};

  /**
   * @param {String} base58key
   * @param {Object.<String, any>} [opts]
   */
  Utils.decode = async function (base58key, opts) {
    //@ts-ignore - wth?
    let DashKeys = window.DashKeys || require("dashkeys");
    //@ts-ignore
    return await DashKeys.decode(base58key, opts);
  };

  /**
   * @param {Uint8Array} keyBytes
   * TODO - pass tprv
   */
  Utils.encodeXPrv = async function (keyBytes) {
    //@ts-ignore - wth?
    let DashKeys = window.DashKeys || require("dashkeys");
    //@ts-ignore
    return await DashKeys.encodeKey(keyBytes, { version: "xprv" });
  };

  /**
   * @param {Uint8Array} keyBytes
   * TODO - pass tpub
   */
  Utils.encodeXPub = async function (keyBytes) {
    //@ts-ignore - wth?
    let DashKeys = window.DashKeys || require("dashkeys");
    //@ts-ignore
    return await DashKeys.encodeKey(keyBytes, { version: "xpub" });
  };
  /** @type {HDKeyTweak} */
  Utils.privateKeyTweakAdd = async function (privateKeyCopy, tweak) {
    let Secp256k1 =
      //@ts-ignore
      window.nobleSecp256k1 || require("@dashincubator/secp256k1");
    let p = Secp256k1.utils._normalizePrivateKey(privateKeyCopy);
    let t = Secp256k1.utils._normalizePrivateKey(tweak);
    return Secp256k1.utils._bigintTo32Bytes(
      Secp256k1.utils.mod(p + t, Secp256k1.CURVE.n),
    );
  };

  /** @type {HDKeyToKey} */
  Utils.publicKeyNormalize = async function (pubBytes) {
    let Secp256k1 =
      //@ts-ignore
      window.nobleSecp256k1 || require("@dashincubator/secp256k1");

    try {
      let point = Secp256k1.Point.fromHex(pubBytes);
      pubBytes = point.toRawBytes(COMPRESSED);
    } catch (e) {
      throw new Error("Invalid public key");
    }

    return pubBytes;
  };

  /** @type {HDKeyTweak} */
  Utils.publicKeyTweakAdd = async function (publicKeyCopy, tweak) {
    let Secp256k1 =
      //@ts-ignore
      window.nobleSecp256k1 || require("@dashincubator/secp256k1");

    let P = Secp256k1.Point.fromHex(publicKeyCopy);
    let t = Secp256k1.utils._normalizePrivateKey(tweak);
    let Q = Secp256k1.Point.BASE.multiplyAndAddUnsafe(P, t, 1n);
    if (!Q) {
      throw new Error("Tweaked point at infinity");
    }
    return Q.toRawBytes(COMPRESSED);
  };

  /** @type {HDHasher} */
  Utils.ripemd160sum = async function (bytes) {
    //@ts-ignore
    let DashKeys = window.DashKeys || require("dashkeys");
    //@ts-ignore
    return await DashKeys.utils.ripemd160sum(bytes);
  };

  /** @type {HDHasher} */
  Utils.sha256sum = async function (bytes) {
    if (!Crypto.subtle) {
      let sha256 = Crypto.createHash("sha256").update(bytes).digest();
      return new Uint8Array(sha256);
    }
    let arrayBuffer = await Crypto.subtle.digest("SHA-256", bytes);
    let hashBytes = new Uint8Array(arrayBuffer);
    return hashBytes;
  };

  /** @type {HDHmac} */
  Utils.sha512hmac = async function (entropy, data) {
    if (!Crypto.subtle) {
      let buf = Crypto.createHmac("sha512", entropy).update(data).digest();
      return new Uint8Array(buf);
    }

    /** @type {"raw"|"pkcs8"|"spki"} */
    let format = "raw";
    let algo = {
      name: "HMAC",
      hash: "SHA-512",
    };
    let extractable = false;
    /** @type {Array<KeyUsage>} */
    let keyUsages = ["sign"];
    let hmacKey = await Crypto.subtle.importKey(
      format,
      entropy,
      algo,
      extractable,
      keyUsages,
    );
    let sig = await Crypto.subtle.sign("HMAC", hmacKey, data);

    return new Uint8Array(sig);
  };

  /** @type {HDSecureErase} */
  Utils.secureErase = function (buf) {
    if (!Crypto.getRandomValues) {
      Crypto.randomBytes(buf.length).copy(buf);
      return;
    }

    Crypto.getRandomValues(buf);
  };

  /** @type {HDKeyToKey} */
  Utils.toPublicKey = async function (privBytes) {
    let Secp256k1 =
      //@ts-ignore
      window.nobleSecp256k1 || require("@dashincubator/secp256k1");

    return Secp256k1.getPublicKey(privBytes, COMPRESSED);
  };

  // "Bitcoin seed"
  let MASTER_SECRET = Uint8Array.from([
    // B     i     t     c     o     i     n   " "     s     e     e     d
    0x42, 0x69, 0x74, 0x63, 0x6f, 0x69, 0x6e, 0x20, 0x73, 0x65, 0x65, 0x64,
  ]);
  let HARDENED_OFFSET = 0x80000000;
  let KEY_SIZE = 33;
  let INDEXED_KEY_SIZE = 4 + KEY_SIZE;
  let XKEY_SIZE = 74;

  // Bitcoin hardcoded by default, can use package `coininfo` for others
  let BITCOIN_VERSIONS = { private: 0x0488ade4, public: 0x0488b21e };

  DashHd.create = function (versions) {
    /** @type {HDKey} */
    let hdkey = {};
    /** @type {Uint8Array?} */
    let _privateKey = null;

    hdkey.versions = versions || BITCOIN_VERSIONS;
    hdkey.depth = 0;
    hdkey.index = 0;
    //hdkey.publicKey = null;
    //hdkey.chainCode = null;
    hdkey.parentFingerprint = 0;

    hdkey.getPrivateKey = function () {
      return _privateKey;
    };
    hdkey.setPrivateKey = function (privBytes) {
      _privateKey = privBytes;
      return null;
    };

    hdkey.getPrivateExtendedKey = async function () {
      if (!_privateKey) {
        return null;
      }

      let key = new Uint8Array(KEY_SIZE);
      key.set([0], 0);
      key.set(_privateKey, 1);
      //@ts-ignore - wth?
      return await Utils.encodeXPrv(serialize(hdkey, key));
    };

    hdkey.getPublicExtendedKey = async function () {
      if (!hdkey.publicKey) {
        throw new Error("Missing public key");
      }

      return await Utils.encodeXPub(serialize(hdkey, hdkey.publicKey));
    };

    hdkey.derive = async function (path) {
      if (path === "m" || path === "M" || path === "m'" || path === "M'") {
        return hdkey;
      }

      let entries = path.split("/");
      let _hdkey = hdkey;
      for (let i = 0; i < entries.length; i += 1) {
        let c = entries[i];
        if (i === 0) {
          assert(/^[mM]{1}/.test(c), 'Path must start with "m" or "M"');
          continue;
        }

        let hardened = c.length > 1 && c[c.length - 1] === "'";
        let childIndex = parseInt(c, 10); // & (HARDENED_OFFSET - 1)
        assert(childIndex < HARDENED_OFFSET, "Invalid index");

        _hdkey = await _hdkey.deriveChild(childIndex, hardened);
      }

      return _hdkey;
    };

    hdkey.deriveChild = async function (index, hardened) {
      for (;;) {
        try {
          //@ts-ignore
          return await hdkey._deriveChild(index, hardened);
        } catch (e) {
          // In essence:
          // if it couldn't produce a public key, just go on the next one
          //
          // More precisely:
          //
          // throw if IL >= n || (privateKey + IL) === 0
          // In case parse256(IL) >= n or ki == 0,
          // one should proceed with the next value for i

          // throw if IL >= n || (g**IL + publicKey) is infinity
          // In case parse256(IL) >= n or Ki is the point at infinity,
          // one should proceed with the next value for i
          index += 1;
        }
      }
    };

    // IMPORTANT: never allow `await` (or other async) between writing to
    // and accessing these! (otherwise the data will be corrupted)
    // (stored here for performance - no allocations or garbage collection)
    let _indexBuffer = new Uint8Array(4);
    let _indexDv = new DataView(_indexBuffer.buffer);

    //@ts-ignore
    hdkey._deriveChild = async function (index, hardened) {
      let seed = new Uint8Array(INDEXED_KEY_SIZE);
      if (hardened) {
        if (!_privateKey) {
          throw new Error("Could not derive hardened child key");
        }
        index += HARDENED_OFFSET;
        seed.set([0], 0);
        seed.set(_privateKey, 1);
      } else {
        seed.set(hdkey.publicKey, 0);
      }
      _indexDv.setUint32(0, index, BUFFER_BE);
      seed.set(_indexBuffer, KEY_SIZE);

      let I = await Utils.sha512hmac(hdkey.chainCode, seed);
      let IL = I.slice(0, 32);
      let IR = I.slice(32);

      let _hdkey = DashHd.create(hdkey.versions);
      _hdkey.depth = hdkey.depth + 1;
      _hdkey.parentFingerprint = await DashHd.fingerprint(hdkey.publicKey);
      _hdkey.index = index;
      _hdkey.chainCode = IR;

      if (_privateKey) {
        let nextPrivKey = await Utils.privateKeyTweakAdd(_privateKey, IL);
        _hdkey.setPrivateKey(nextPrivKey);
        _hdkey.publicKey = await Utils.toPublicKey(nextPrivKey);
      } else {
        _hdkey.publicKey = await Utils.publicKeyTweakAdd(hdkey.publicKey, IL);
      }

      return _hdkey;
    };

    hdkey.wipePrivateData = function () {
      if (_privateKey) {
        Utils.secureErase(_privateKey);
      }
      _privateKey = null;
      return hdkey;
    };

    return hdkey;
  };

  /** @type {HDFingerprint} */
  DashHd.fingerprint = async function (pubBytes) {
    if (!pubBytes) {
      throw new Error("Public key has not been set");
    }

    /*
     * Note: this *happens* to use the same algorithm
     * as many toPkh() implementations but, semantically,
     * this is NOT toPkh() - it has a different purpose.
     * Furthermore, fingerprint() may change independently of toPkh().
     */
    let sha = await Utils.sha256sum(pubBytes);
    let identifier = await Utils.ripemd160sum(sha);
    let i32be = readUInt32BE(identifier, 0);
    return i32be;
  };

  /**
   * @param {Uint8Array} u8 - a "web" JS buffer
   * @param {Number} offset - where to start reading
   * @returns {Number} - a 0-shifted (uint) JS Number
   */
  function readUInt32BE(u8, offset) {
    let dv = new DataView(u8.buffer);
    // will read offset + 4 bytes (32-bit uint)
    let n = dv.getUint32(offset, BUFFER_BE);
    return n;
  }

  DashHd.fromMasterSeed = async function (seedBuffer, versions) {
    let I = await Utils.sha512hmac(MASTER_SECRET, seedBuffer);
    let IL = I.subarray(0, 32);
    let IR = I.subarray(32);

    let hdkey = DashHd.create(versions);
    hdkey.chainCode = IR;
    await hdkey.setPrivateKey(IL);
    hdkey.publicKey = await Utils.toPublicKey(IL);

    return hdkey;
  };

  DashHd.fromExtendedKey = async function (
    base58key,
    versions,
    skipVerification,
  ) {
    // => version(4) || depth(1) || fingerprint(4) || index(4) || chain(32) || key(33)
    versions = versions || BITCOIN_VERSIONS;
    skipVerification = skipVerification || false;
    let hdkey = DashHd.create(versions);

    //@ts-ignore - wth?
    let keyInfo = await Utils.decode(base58key);
    let keyBytes = DashKeys.utils.hexToBytes(keyInfo.xprv || keyInfo.xpub);
    let keyDv = new DataView(keyBytes.buffer, 0, keyBytes.byteLength);

    //let version = keyDv.getUint32(0, BUFFER_BE);
    // TODO tprv, tpub
    let version = parseInt(keyInfo.version, 16);
    assert(
      version === versions.private || version === versions.public,
      "Version mismatch: does not match private or public",
    );

    hdkey.depth = keyDv.getUint8(0);
    hdkey.parentFingerprint = keyDv.getUint32(1, BUFFER_BE);
    hdkey.index = keyDv.getUint32(5, BUFFER_BE);
    hdkey.chainCode = keyBytes.subarray(9, 41);

    let key = keyBytes.subarray(41);
    if (keyDv.getUint8(41) === 0) {
      // private
      assert(
        version === versions.private,
        "Version mismatch: version does not match private",
      );
      let privBytes = key.subarray(1); // cut off first 0x0 byte
      await hdkey.setPrivateKey(privBytes);
      hdkey.publicKey = await Utils.toPublicKey(privBytes);
    } else {
      assert(
        version === versions.public,
        "Version mismatch: version does not match public",
      );
      assert(
        key.length === 33 || key.length === 65,
        "Public key must be 33 or 65 bytes.",
      );
      hdkey.publicKey = key;
      if (!skipVerification) {
        hdkey.publicKey = await Utils.publicKeyNormalize(hdkey.publicKey);
      }
    }

    return hdkey;
  };

  /**
   * @param {Boolean} assertion
   * @param {String} message
   */
  function assert(assertion, message) {
    if (!assertion) {
      throw new Error(message);
    }
  }

  /**
   * @param {HDKey} hdkey - TODO attach to hdkey
   * @param {Uint8Array} key
   */
  function serialize(hdkey, key) {
    // => version(4) || depth(1) || fingerprint(4) || index(4) || chain(32) || key(33)
    let xkey = new Uint8Array(XKEY_SIZE);
    let xkeyDv = new DataView(xkey.buffer);

    xkeyDv.setUint8(0, hdkey.depth);

    let fingerprint = (hdkey.depth && hdkey.parentFingerprint) || 0x00000000;
    xkeyDv.setUint32(1, fingerprint, BUFFER_BE);
    xkeyDv.setUint32(5, hdkey.index, BUFFER_BE);

    xkey.set(hdkey.chainCode, 9);
    xkey.set(key, 41);

    return xkey;
  }

  DashHd.HARDENED_OFFSET = HARDENED_OFFSET;
})(("object" === typeof window && window) || {}, DashHd);
if ("object" === typeof module) {
  module.exports = DashHd;
}

// Type Definitions

/**
 * @typedef HDVersions
 * @prop {Number} private - 32-bit (4-byte) int (encodes to 'xprv' in base58)
 * @prop {Number} public - 32-bit (4-byte) int (encodes to 'xpub' in base58)
 */

// Function Definitions

/**
 * @callback HDDeriveChild
 * @param {Number} index - includes HARDENED_OFFSET, if applicable
 * @param {Boolean} hardened
 * returns {Promise<HDKey>}
 */

/**
 * @callback HDDerivePath
 * @param {String} path
 * returns {Promise<HDKey>}
 */

/**
 * @callback HDFingerprint
 * @param {Uint8Array} pubBytes - Public Key
 * @returns {Number}
 */

/**
 * @callback HDFromXKey
 * @param {String} base58key - base58check-encoded xkey
 * @param {HDVersions} [versions]
 * @param {Boolean} [skipVerification]
 * returns {Promise<HDKey>}
 */

/**
 * @callback HDFromSeed
 * @param {Uint8Array} seedBuffer
 * @param {HDVersions} [versions]
 * @returns {Promise<HDKey>}
 */

/**
 * @callback HDGetBuffer
 * @returns {Uint8Array}
 */

/**
 * @callback HDGetString
 * @returns {Promise<String>}
 */

/**
 * @callback HDHasher
 * @param {Uint8Array} bytes
 * @returns {Promise<Uint8Array>} - hash Uint8Array
 */

/**
 * @callback HDHmac
 * @param {Uint8Array} entropy - secret, salt, or key
 * @param {Uint8Array} data - message
 * @returns {Promise<Uint8Array>} - hash Uint8Array
 */

/**
 * @callback HDKeyToKey
 * @param {Uint8Array} keyBytes - Key Uint8Array
 * @returns {Promise<Uint8Array>} - Public Key Uint8Array
 */

/**
 * @callback HDKeyTweak
 * @param {Uint8Array} keyBytes - Key Uint8Array
 * @param {Uint8Array} tweakBytes - Tweak Uint8Array
 * @returns {Promise<Uint8Array>} - Public Key Uint8Array
 */

/**
 * @callback HDMaybeGetBuffer
 * @returns {Uint8Array?}
 */

/**
 * @callback HDMaybeGetString
 * @returns {Promise<String?>}
 */

/**
 * @callback HDSecureErase
 * @param {Uint8Array} buf
 * @returns {void}
 */

/**
 * @callback HDSetBuffer
 * @param {Uint8Array} buf
 */

/**
 * @callback HDWipePrivates
 */
